ls1intum/Artemis
# [Feature] Hidden/private tests

### Checklist

### Motivation and Context
Closes #2487.
Motivation described there.

### Description
I implemented the solution as outlined in #2487 with `Feedback` and `ProgrammingExerciseTestCase` both receiving a `visibility`-Attribute.
This means that all `Feedbacks` received are stored and only filtered when queried depending on the individual user.

Additionally, as visibility for `Feedbacks` marked with after due date depends only on the Due Date of the exercise itself and no longer on *Run Tests After Due Date* it is always possible to select *AfterDueDate* as option for test cases.
To update the score after they become visible, now a `updateAllResults` is triggered for all student participations after the due date has passed.
This only recalculates the result string and score based on values already stored and does not require any Bamboo/Jenkins jobs.

### Steps for Testing
There aren’t really specific steps as this affects a lot of the programming exercise grading workflow, but some things you should look out for:
- This change should only affect programming exercises.
- The only UI-Page that has changed is the programming exercise grading configuration page where the column *After Due Date* has been changed to *Visibility*.
    - Now *AfterDueDate* should always be selectable in the dropdown independent of the *Run Tests After Due Date* option for the exercise.
- Test cases marked as *AfterDueDate* should
    - be invisible to students (e.g. in the Feedback list, but other possible places as well) before the due date has passed,
    - become visible to students after the due date has passed,
    - always be visible in the feedback list when teaching assistants/instructors look at a student’s participation in the participations view and show up for the statistics on the *Configure Grading* page in the column *Passed %* (if they are actually executed on the CI server),
    - be considered in the result string (*x of y passed*) only after the due date has passed.
- Test cases marked as *Never* visible should
    - always be invisible to students,
    - be visible in the feedback list when teaching assistants/instructors look at a student’s solution in the participations view and when manually assessing a result,
    - only be considered in the result string (*x of y passed*) for the `Template` and `Solution` repos; never for student repositories.
- When manually assessing a result
    - feedback for test cases marked as *Never* visible should be visible to the assessor.
    - students should not see the automatic feedback for a *Never*-testcase after a manual result has been created.

### Test Coverage
| Class | Instruction/Statement Coverage |
| ----- | ----------- |
| domain/Result.java | 95% (unchanged, all new code covered) |
| service/ProgrammingExerciseGradingService.java | 95% (unchanged, all new code covered) |
| service/WebsocketMessagingService.java | 75% (unchanged, all new code covered) |
| service/scheduled/ProgrammingExerciseScheduleService.java | 74% → 75% |
| web/rest/ProgrammingExerciseParticipationResource.java | 90% → 91% |
| web/rest/ResultResource.java | 87% (unchanged, all new code covered) |
| programming-exercise-configure-grading.component.ts | 88% (unchanged) |

### Screenshots
<details>
  <summary>Click to expand!</summary>

Configure Grading Page (`testMergeSort`: Hidden, `testBubbleSort`: After Due Date), it is shown by the green indicator that a student successfully implemented those two tests:
![grading_page_with_student_submission](https://user-images.githubusercontent.com/64250573/105474047-1dd4b700-5c9e-11eb-98a9-1dd128fb14c8.png)
Instructor View of the Solution Repository (hidden tests shown):
![solution_view](https://user-images.githubusercontent.com/64250573/105474050-1e6d4d80-5c9e-11eb-9539-ee626d4494d2.png)
Student view before due date (hidden tests and tests after due date not shown):
![student_view_before_due_date](https://user-images.githubusercontent.com/64250573/105474052-1f9e7a80-5c9e-11eb-8093-54b3cbdf1a3c.png)
Student view after due date (hidden tests not shown but tests after due date are visible):
![student_view_after_due_date](https://user-images.githubusercontent.com/64250573/105474051-1f05e400-5c9e-11eb-92c9-62b8a8f1593e.png)

</details>
